### PR TITLE
New Instrument: MKS937B vacuum gauge controller

### PR DESCRIPTION
**Pull request by dkriegner**
This implements a device of a new vendor: [MKS Instruments 937B vacuum gauge controller](https://www.mks.com/f/937b-combination-vacuum-gauge-controller)

Connection to the device is made through an RS232/RS485 serial connection. 
The communication protocol of this device is as follows:

Query: `@<aaa><Command>?;FF` with the response `@<aaa>ACK<Response>;FF`
Set command: `@<aaa><Command>!<parameter>;FF` with the response `@<aaa>ACK<Response>;FF`

`<aaa>` is an address from `001` to `254`

I did not see a way to implement this without overloading VISAAdapter since the terminator `';FF'` is invalid in pyvisa as `read_terminator` (`ValueError: ambiguous ending in termination characters`). But if you see a more elegant solution, please let me know.

Using the code in this pull request, the device can be initialized via:
```
import pymeasure.instruments
gauge = pymeasure.instruments.mksinst.MKS937B('ASRL1::INSTR')  # or any other serial device
```

Edit:
I should also point out that I propose some changes to `Instrument` to allow map values to be used also in case multiple values are returned by an instrument. I guess this needs particular attention during the code review. Maybe even some test case has to be added to test this functionality? (How would this be done?)

### CODE REVIEW

**Comment by bmoneke** [comment]
We're implementing a testing suite, with which you can test the communication protocol of your device. PR #634 contains the current work in progress.
The tests do work, we're writing the documentation (first version is already done, so look at [adding_instruments](docs/dev/adding_instruments.rst) in the protocol-test branch.

Instead of writing your own adapter, you could add reading those two bytes to the instruments read/values method.
I think that would be more in line with the current plan to have all instrument related stuff in the instrument and not the adapter itself. I'm not a maintainer, however, so wait for some official answer.

File: pymeasure/instruments/mksinst/mks937b.py
Comment (bmoneke) [inline pymeasure/instruments/mksinst/mks937b.py:90]: You want to include the device address in the command sent to the device, don't you?
Alternatively, you can rewrite the instrument write/values, to do that for you. The values/read methods could check for those two additional "F" chars, as well.

That way you do not make all the properties dynamic.
```suggestion

    def values(self, command, **kwargs):
        ret = super().values(self.command_process(command), **kwargs)
        self._check_termination()
        return ret

    def _check_termination(self):
        """Check the read termination to correspond to the protocol."""
        t = self.adapter.read_bytes(2)  # read extra termination chars 'FF'
        if t != b'FF':
            raise ValueError(f"unexpected termination string received {t}")
```

**Comment by dkriegner** [comment]
> We're implementing a testing suite, with which you can test the communication protocol of your device. PR #634 contains the current work in progress. The tests do work, we're writing the documentation (first version is already done, so look at [adding_instruments](docs/dev/adding_instruments.rst) in the protocol-test branch.

I have not yet had the time to look into that.
 
> Instead of writing your own adapter, you could add reading those two bytes to the instruments read/values method. I think that would be more in line with the current plan to have all instrument related stuff in the instrument and not the adapter itself. I'm not a maintainer, however, so wait for some official answer.

I followed your advise and revised the code to avoid the definition of an extra adapter. I am not sure if I should not also overload the Instrument.ask method. Its not needed for the present code, but for the user it could be best to provide a consistent interface where ask/write/values performs the adding of the address part in the command.

File: pymeasure/instruments/mksinst/mks937b.py
Comment (dkriegner) [reply pymeasure/instruments/mksinst/mks937b.py]: This method depends on self (see self._re_response_value, self.re_fullresponse) which need to be instance variables since they depend on the device address. They could be made generic regular expressions, but I find it best to check it with the actual device address and not just with any 3 digit number. If this device would be used on a real bus connection (RS485) it is likely best in this way.

File: pymeasure/instruments/mksinst/mks937b.py
Comment (dkriegner) [reply pymeasure/instruments/mksinst/mks937b.py]: `check_errors` has to be exposed since this is how it used inside the `Instrument.control` properties code. The other one I mark private.

File: pymeasure/instruments/mksinst/mks937b.py
Comment (bmoneke) [reply pymeasure/instruments/mksinst/mks937b.py]: Sorry, I missed that self, so you're fine.

**Comment by bmoneke** [comment]
We plan to make Instrument.ask call write+read of the instrument instead of calling Adapter.ask, because we want to have the whole "protocol logic" in these two methods: write and read.
If you're concerned somebody might use ask, implement it as self.write+self.read.

**Comment by dkriegner** [comment]
> We plan to make Instrument.ask call write+read of the instrument instead of calling Adapter.ask, because we want to have the whole "protocol logic" in these two methods: write and read. If you're concerned somebody might use ask, implement it as self.write+self.read.

This sounds like a sensible change. I already had to generate an Adapter because the Instrument read/write are effectively bypassed. I think it would make sense to do this on the Instrument level and not in every device. I think this can wait for a central solution of this issue.

> 
> It would be great, if you added tests for the expanded Instrument.control to the corresponding [test file](https://github.com/pymeasure/pymeasure/blob/master/tests/instruments/test_instrument.py).

will do. please wait with the merge for my tests

**Comment by dkriegner** [comment]
I have added a test to the mentioned file. 

Since the feature of mapping was only implemented in `Instrument.measurement` I limited the test to the a `measurement` property. In order to extend the mapping of multiple values also to the `setting` or `control` properties I think more extensive changes are needed which need to be discussed.

- validators would need to handle iterable objects
- either the default separator (which now seems only available in `Adapter.values`) would need to be accessible to `Instrument.control.fset` in order that the setting string can be generated. 
Or 
- the set string would need to have multiple variable placeholders (which for really long arrays would be likely impractical)

It seems to me this would be beyond this scope of this PR. From my point of view this Instrument implementation would be ready to be merged.

**Comment by bilderbuchi** [comment]
Looks interesting, but I'll need more time until I get around for a review, summer time is crazy :D (also, I'd like to get #634 in, first)

**Comment by bilderbuchi** [comment]
> Looks interesting, but I'll need more time until I get around for a review, summer time is crazy :D (also, I'd like to get #634 in, first)

just FYI, #634 has been merged now, so [protocol tests are possible now](https://pymeasure.readthedocs.io/en/latest/dev/adding_instruments.html#protocol-tests).

**Comment by bilderbuchi** [comment]
> I did not see a way to implement this without overloading VISAAdapter since the terminator `';FF'` is invalid in pyvisa as `read_terminator` (`ValueError: ambiguous ending in termination characters`). But if you see a more elegant solution, please let me know.

yeah, that one will be challenging to get around.

**Review by bilderbuchi — CHANGES_REQUESTED** [review]
First round of review, without the changes to map multiple values -- I might have a better approach, but I will need some more thought.

File: pymeasure/instruments/instrument.py
Comment (bilderbuchi) [inline pymeasure/instruments/instrument.py:397]: What I don't like is that this duplicates a lot of code and logic. I have come up with the following better alternative, which passes your test but does not duplicate the code. 
```
        def fget(self,
                 get_command=get_command,
                 values=values,
                 map_values=map_values,
                 get_process=get_process,
                 command_process=command_process,
                 check_get_errors=check_get_errors,
                 ):
            if get_command is None:
                raise LookupError("Instrument property can not be read.")
            vals = self.values(command_process(get_command), **kwargs)
            if check_get_errors:
                self.check_errors()
            vals = [get_process(v) for v in vals]  # TODO: should we run get_process on all elements or the whole list?
            if map_values:
                ret = []
                for value in vals:
                    if isinstance(values, (list, tuple, range)):
                        ret.append(values[int(value)])
                    elif isinstance(values, dict):
                        for k, v in values.items():
                            if v == value:
                                ret.append(k)
                                break
                        else:  # for-loop ran completely
                            raise KeyError(f"Value {value} not found in mapped values")
                    else:
                        raise ValueError(
                            'Values of type `{}` are not allowed '
                            'for Instrument.control'.format(type(values))
                        )
            else:
                ret = vals
            if len(ret) == 1:
                return ret[0]
            else:
                return ret
```
(basically, I am delaying the distinction between 1 and multiple value elements to later and do mapping on all elements)

However, I am not yet 100% confident that there are not further implications that I did not consider. 
* Also, I have not yet thought through at which point get_process should apply. 
* I'm not sure about mapping now becoming asymmetric, as mapping lists of items now only applies on get. Does it even make sense on set?

It would be good to have the extended amount of tests from #665 (will merge soon) before merging this.

ping @bmoneke what do you think about the impact of this?

**Review by bilderbuchi — CHANGES_REQUESTED** [review]
Some thoughts on the values mapping change.

File: pymeasure/instruments/mksinst/mks937b.py
Comment (bilderbuchi) [inline pymeasure/instruments/mksinst/mks937b.py:107]: @bmoneke that approach might be more challenging to test with our current testing setup (either ProtocolAdapter or the Mock when checking all instruments). I guess we will see on the next CI run :D

File: pymeasure/instruments/mksinst/mks937b.py
Comment (bilderbuchi) [reply pymeasure/instruments/mksinst/mks937b.py:105]: We could probably to an instrument-specific test that this all works correctly using pyvisa-sim like [here](https://github.com/pyvisa/pyvisa/issues/120#issuecomment-74980018), where we can create simple returns, and include all termination characters. not sure if the eom entry can accept `;FF`, though.

File: pymeasure/instruments/mksinst/mks937b.py
Comment (bmoneke) [reply pymeasure/instruments/mksinst/mks937b.py:107]: You have to make the communication pairs longer, but it should work.
E.g.
comm_pairs = [(b"writing something", b"the response"), (None, b"FF")]

Remember, that you must not use the termination_characters defined at adapter level (i.e. ";FF" for writing and ";" for reading).

File: pymeasure/instruments/instrument.py
Comment (bmoneke) [reply pymeasure/instruments/instrument.py:397]: I'd say make `get_process(vals)`, because it gives flexibility without loosing possibility:

- If you want to modify every entry of a list, get_process can take that into account. e.g. `lambda vals: [val**2 for val in vals]`.
- If you want to take the list as a whole, e.g. get the average or standard deviation or sum or min, max or a specific value (e.g. the first one) etc., you can do it as well.

Comment (bmoneke) [reply pymeasure/instruments/instrument.py:397]: I just noticed problem with my suggestion:
If you had `get_process = lambda v: 2*v` before, you need to work with the list now.

I do not know by heart how many instruments implement get_process.

On the other hand, applying the get process to every single value should not brake any present code.

Comment (bilderbuchi) [reply pymeasure/instruments/instrument.py:397]: Good point on the flexibility!

Backwards compatibility is a good point (get_process is used a lot)! 
Until now, get_process is acting on the string entry if vals was length 1, and on the list if length >1 (which may be less than ideal in itself). We could preserve that behaviour with some small code tweaks.

File: pymeasure/instruments/mksinst/mks937b.py
Comment (dkriegner) [reply pymeasure/instruments/mksinst/mks937b.py]: used the suggested code (and also fixed the docstring) accordingly

File: pymeasure/instruments/mksinst/mks937b.py
Comment (dkriegner) [reply pymeasure/instruments/mksinst/mks937b.py]: thanks for the suggestion. I have improved the code accordingly

File: pymeasure/instruments/mksinst/mks937b.py
Comment (dkriegner) [reply pymeasure/instruments/mksinst/mks937b.py]: I went with your suggestion and simplified the code

File: pymeasure/instruments/instrument.py
Comment (dkriegner) [reply pymeasure/instruments/instrument.py]: I used the code suggestion from above. Its true that is easier like you propose it. It seems to work for my device. 
I did not understand from your discussion if you still want to handle get_process differently?

File: pymeasure/instruments/mksinst/mks937b.py
Comment (dkriegner) [reply pymeasure/instruments/mksinst/mks937b.py:111]: I tried to understand how the protocol test could work for this device but did not succeed. From the [documentation](https://github.com/pymeasure/pymeasure/blob/master/docs/dev/adding_instruments.rst) I assumed I can write something like:

```
import pytest

from pymeasure.test import expected_protocol
from pymeasure.instruments.mksinst.mks937b import MKS937B

def test_pressure():
    """Verify the communication of the voltage getter."""
    with expected_protocol(
        MKS937B,
        [("@253PR1?", "@253ACK1.10e-9"),
         (None, b"FF")],
    ) as inst:
        assert inst.pressure1 == 1.1e-9
```

but this failed with:
` AssertionError: assert '@253ACK1.10e-9' == 1.1e-09`

Can you give me a hint on how this is done for this instrument? I guess it makes sense I write few basic tests for this to be included in this PR

**Comment by dkriegner** [comment]
I can not see how the latest CI run fail is related to this pull request... If I should merge main (which might be needed anyway to have the latest protocol tests included) or perform any other changes please let me know

File: pymeasure/instruments/mksinst/mks937b.py
Comment (bmoneke) [reply pymeasure/instruments/mksinst/mks937b.py:111]: The test looks right and I guess it does its job, exposing coding errors 😀: `inst.pressure1` returns the string the device responded and probably you made somewhere in your code a mistake (or values does not use the `preprocess_reply`).
Maybe your `_extract_reply` does not work as expected. I'd write a test for that method as well.

Comment (dkriegner) [reply pymeasure/instruments/mksinst/mks937b.py:111]: Not sure i can agree here. testing the same with the hardware seems to work
```
>>> import pymeasure.instruments
>>> inst = pymeasure.instruments.mksinst.MKS937B("ASRL3::INSTR")
>>> inst.pressure1
1.6e-08
>>>
```
So with the hardware all functions seem to be called as I intend it. Something seems different in the test case?

Comment (dkriegner) [reply pymeasure/instruments/mksinst/mks937b.py:111]: As extra info here the raw hardware communication for the same command:
```
>>> import pyvisa
>>> rm = pyvisa.ResourceManager()
>>> inst = rm.open_resource("ASRL3::INSTR")
>>> inst.write("@253PR1?;FF")
13
>>> inst.read_bytes(18)
b'@253ACK1.50E-08;FF'
>>>
```

Comment (bmoneke) [reply pymeasure/instruments/mksinst/mks937b.py:111]: I found the problem: The ProtocolAdapter does not get the global preprocess_reply configuration, as only a newly created VISAAdapter gets that information. (I'm reworking the adapters, so that's going to change in the future).

As a workaround you can add `inst.adapter.preprocess_reply = inst._extract_reply` directly after the colon in the test. That way you add the preprocess to the adapter and the test succeeds.
In theory you can hand parameters to the ProtocolAdapter if you define them after the communication pairs. In you case, that does not work, as the preprocess_reply is a method of the instrument instance.

Comment (dkriegner) [reply pymeasure/instruments/mksinst/mks937b.py:111]: thanks with the workaround the tests succeed. I have added some to test the different types of properties of this new device.

**Comment by dkriegner** [comment]
> I can not see how the latest CI run fail is related to this pull request... If I should merge main (which might be needed anyway to have the latest protocol tests included) or perform any other changes please let me know

after merging master and adding the protocol tests no more error appaers...

@bilderbuchi: The MKS937B code is ready for merging from my point of view. You certainly still want to check the changes in the Instrument. measurement getter method, but the tests present on the master branch seem to succeed

File: pymeasure/instruments/mksinst/mks937b.py
Comment (bmoneke) [reply pymeasure/instruments/mksinst/mks937b.py:111]: Thanks for the tests. It makes future maintenance easier. You have some style errors (one instead of two newlines etc.), see the file with the comments made by CI.

Comment (bilderbuchi) [reply pymeasure/instruments/mksinst/mks937b.py:111]: Great troubleshooting, folks! Thanks for supplying complete information; thanks for helping out getting a good test running! "This is the way!" :D

File: pymeasure/instruments/mksinst/mks937b.py
Comment (bilderbuchi) [reply pymeasure/instruments/mksinst/mks937b.py:109]: The added test takes care of this already.

File: pymeasure/instruments/instrument.py
Comment (bilderbuchi) [reply pymeasure/instruments/instrument.py]: > I did not understand from your discussion if you still want to handle get_process differently?

> Event though it does not offer that much flexibility (e.g. averaging), I think your implementation is good, due to backwards compatibility.

> On the other hand, applying the get process to every single value should not brake any present code.

I think it's the other way round, we have changed behaviour of `get_process` when multiple elements are present:

> Until now, get_process is acting on the string entry if vals was length 1, and on the list if length >1 (which may be less than ideal in itself). We could preserve that behaviour with some small code tweaks.

I'll think on this, come up with a test and will suggest a tweak of that part of the code.

**Review by bilderbuchi — CHANGES_REQUESTED** [review]
We are nearly there, I think the main thing that's left is the `get_process` behaviour.

File: pymeasure/instruments/instrument.py
Comment (bilderbuchi) [reply pymeasure/instruments/instrument.py]: OTOH, the docstring of `get_process` is "A function that take a value and allows processing before value mapping, returning the processed value", suggesting it acts on single values. 

In addition, we have the `preprocess_reply` that acts on the whole message (before it runs through `values()`, though)

Also, having  `get_process` always act on single elements simplifies the situation when the returned list can have a variable number of elements (but I don't know how frequent that is).

File: pymeasure/instruments/mksinst/mks937b.py
Comment (dkriegner) [reply pymeasure/instruments/mksinst/mks937b.py:101]: my bad. fixed

**Comment by dkriegner** [comment]
> We are nearly there, I think the main thing that's left is the `get_process` behaviour.

thanks for the review. I think I will leave the `get_process` for you since I do not have the bigger picture what is wanted for this particular function. Personally I think applying it to a single value makes more sense since as you wrote `preprocess_reply` covers the full message already.

If more changes are needed from me just let me know.

File: pymeasure/instruments/instrument.py
Comment (bilderbuchi) [reply pymeasure/instruments/instrument.py]: @msmttchr @bmoneke what do you think the behaviour of `get_process()` with multiple values should be? Act on the whole list (like now, would make code slightly more complex, but is more flexible), or individual elements (fits smoother into the control flow, more consistent with single-element list use).
I'm on the fence, and my dice are a bad basis for decisions here. :grin:

**Comment by dkriegner** [comment]
something in the latest master branch breaks this code.

It can be fixed by this patch:
```
--- a/pymeasure/instruments/mksinst/mks937b.py
+++ b/pymeasure/instruments/mksinst/mks937b.py
@@ -71,7 +71,6 @@ class MKS937B(Instrument):
     """

     def __init__(self, adapter, address=253, **kwargs):
-        kwargs.setdefault("preprocess_reply", self._extract_reply)
         kwargs.setdefault("write_termination", ";FF")
         kwargs.setdefault("read_termination", ";")  # in reality its ";FF"
         # which is, however, invalid for pyvisa. Therefore extra bytes have to
@@ -108,7 +107,7 @@ class MKS937B(Instrument):
         """
         Check the read termination to correspond to the protocol
         """
-        t = self.adapter.read_bytes(2)  # read extra termination chars 'FF'
+        t = super().read_bytes(2)  # read extra termination chars 'FF'
         if t != b'FF':
             raise ValueError(f"unexpected termination string received {t}")

@@ -137,8 +136,8 @@ class MKS937B(Instrument):

         :param command: command string to be sent to the instrument
         """
+        kwargs.setdefault("preprocess_reply", self._extract_reply)
         ret = super().values(self._prepend_address(command), **kwargs)
-        self._check_extra_termination()
         return ret

     def check_errors(self):
```

Is this expected to be needed due to recent changes? And is it the best way (should I push a commit with these changes and rebase on master)?

**Comment by bmoneke** [comment]
In the recent change, we moved all those "special things" from a preprocess_reply into the read / write methods of the instrument itself. And `values` calls `write` and `read` (via ask), therefore you can (as you already did) remove the check from your `values` (last section of your changes).

If `_extract_reply` is useful for every read (I'd guess so), add it to `read` itself (and you can delete your `values` implementation), otherwise you can do it, as it is right now.

Another thing: You can include "_prepend_address" already in `write`, as every writing action (values, ask, or directly 'write') goes through this method.

I'd say, merging master is fine, and rebase is a better.

**Comment by dkriegner** [comment]
I agree that these recent changes make a lot of sense. I now removed my overloading of `values` completely and simplified the rest as recommended. I would keep `_extract_reply` and `_prepend_address` in own methods which makes them better documented.

**Comment by bilderbuchi** [comment]
I think we need some more thought on how we want `get_process` and mapping multiple values to behave (see the unresolved conversation). It's a pity that this drags out things, but changes to pymeasures core logic are always much more hairy than simply adding another instrument. I ask for your patience.

File: pymeasure/instruments/instrument.py
Comment (bmoneke) [reply pymeasure/instruments/instrument.py]: Sorry for not coming back. It was hard for me to decide, what to do. I definitively like the idea of doing some fancy method on the returned values list (e.g. sum and standard deviation...), but that would require to have a more complex `get_process = lambda v: 5 * v[0]` instead of just multiplying with some number.

Whenever you need complex mathematics on the return values (I'd say rarely), you can use a python property. In the many cases of a single return value, we should have a more simple `get_process`.

Therefore I propose that `get_process` acts on every single value.

Comment (bilderbuchi) [reply pymeasure/instruments/instrument.py]: Thank you! I opened #740 for nailing down the current behaviour, integrating the PR, and writing docs and changelog for this change. Let us continue the discussion there.

**Comment by bilderbuchi** [comment]
Thank you, we consider your contribution complete. We will hash out the remaining details in #740, where I have also integrated your branch. This PR will close when we merge #740.

**Comment by bilderbuchi** [comment]
Superseded by #772 (and later #740 for the sensor types properties)

**Pull request closed without merging**
